Stop using Vector::unsafeAppendWithoutCapacityCheck() in JSObject::putOwnDataPropertyBatching()

cdumez opened the pull request:
#### b12bca52a799310cd7338160834a167d4ca67252
<pre>
Stop using Vector::unsafeAppendWithoutCapacityCheck() in JSObject::putOwnDataPropertyBatching()
<a href="https://bugs.webkit.org/show_bug.cgi?id=265084">https://bugs.webkit.org/show_bug.cgi?id=265084</a>

Reviewed by Darin Adler.

Update the `Vector&lt;T&gt;(size_t, Functor)` to abort appending if the Functor
returns a std::optional&lt;T&gt; type that is std::nullopt.

Adopt it in JSObject::putOwnDataPropertyBatching() to avoid calling
Vector::unsafeAppendWithoutCapacityCheck() explicitly.

* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::putOwnDataPropertyBatching):
* Source/WTF/wtf/Vector.h:
(WTF::Vector::Vector):

Canonical link: <a href="https://commits.webkit.org/270980@main">https://commits.webkit.org/270980@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c337758e1c3642b8d4b068450d603073b4066dc2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26919 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5535 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28155 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29130 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24609 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7388 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2927 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24491 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27181 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4362 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23112 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3821 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3885 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24110 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29765 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23460 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24567 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24515 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30106 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/26105 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3898 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2100 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28017 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5358 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33557 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6479 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4365 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7259 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4268 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -714,8 +714,17 @@ class Vector : private VectorBuffer<T, inlineCapacity, Malloc> {
 
-        for (size_t i = 0; i < size; ++i)
-            unsafeAppendWithoutCapacityCheck(valueGenerator(i));
+        if constexpr (std::is_same_v<std::invoke_result_t<Functor, size_t>, std::optional<T>>) {
```
Should there be a `shrinkToFit` somewhere? Seems unclear if we should keep extra capacity indefinitely. I’d expect this to be the same as `compactMap` in that respect.

cdumez replied:
This is something that bugs me with compactMap(). Shouldn't we leave this up to the caller? If the caller wants to store the vector somewhere for an extended period of time, they can call shrinkToFit().

In this instance, the only call site doesn't really want to call shrinkToFit() as the vector is on the stack and doesn't live long (the previous code didn't shrinkToFit()). If I force a shrinkToFit() in the constructor then this call site will pay a cost unnecessarily.

cdumez replied:
Given that was would be your preference? My plan was to not call it and leave it up to the caller (The current caller doesn't call it because it would only hurt performance without benefit AFAIK). Alternatively, we could add a parameter to decide whether or not to call shrinkToFit() and make it the default behavior (optional parameter).

darinadler replied:
Sure that makes sense.

Getting this resizing optimal is tricky and I wish it wasn't. In older code the most common mistake was keeping a vector around for a long time with excess capacity and @smfr tried to change our idioms to decrease the chance of that mistake.

But clearly **this** call site doesn't want a `shrinkToFit`. I'm just not sure how we can make it easy to get right. If we are keeping a vector around for a while we want it right-sized and I can easily imagine someone using this constructor for that kind of situation in the future. For now I guess it's OK to be inconsistent with `compactMap` while we think about the best future idiom.

cdumez committed:
Stop using Vector::unsafeAppendWithoutCapacityCheck() in JSObject::putOwnDataPropertyBatching()
https://bugs.webkit.org/show_bug.cgi?id=265084

Reviewed by Darin Adler.

Update the `Vector<T>(size_t, Functor)` to abort appending if the Functor
returns a std::optional<T> type that is std::nullopt.

Adopt it in JSObject::putOwnDataPropertyBatching() to avoid calling
Vector::unsafeAppendWithoutCapacityCheck() explicitly.

* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::putOwnDataPropertyBatching):
* Source/WTF/wtf/Vector.h:
(WTF::Vector::Vector):

Canonical link: https://commits.webkit.org/270980@main
